OPHD/RandomNumberGenerator.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <type_traits>
+#include <stdexcept>
+#include <random>
+
+
+class RandomNumberGenerator
+{
+public:
+	RandomNumberGenerator() : generator(randomDevice()) {};
+
+	template<typename T, std::enable_if_t<std::is_arithmetic_v<T>>>
+	T generate(T min, T max)
```

Comment:
This is one way to fix that problem:
```cpp
	template<typename T, typename = std::enable_if_t<std::is_arithmetic_v<T>>>
	T generate(T min, T max)
```

I may have forgotten to include this in my previous review. I had made the change temporarily, but thought I had removed it afterwards, which is why I didn't include it as part of the review.

I had also experimented with using `std::enable_if_t` on the return type, rather than a template parameter. Normally that's what I've done in the past. I used to avoid extra template type parameters, since that means the user could potentially specify that extra parameter at the point of call:

It weird, but valid to include an explicit type for the second type parameter, such as `void`, or `int`, using the above syntax for the function definition.
```cpp
randomNumber.generate<int, void>(0, 1000)
```

If you want to avoid that possibility, you can place `std::enable_if_t` on the return type instead:
```suggestion
	template<typename T>
	std::enable_if_t<std::is_arithmetic_v<T>, T>
	generate(T min, T max)
```

Note that you must be explicit with the second type parameter to `std::enable_if_t`, since the default is `void`, and we want the return type to be `T` rather than `void`.

Reply:
I figured specifying `template<typename T, std::enable_if_t<std::is_arithmetic_v<T>>>` covered all instances of T in the function declaration. Based on what you are saying, I'm thinking that this is not the case and it only covers the 2 arguments min and max.

I remember reading long ago that some of the template material was embedded in the function declaration. I was wondering if that affected the syntax here?

Reply:
Using `std::enable_if` is a bit weird. It basically works by creating invalid syntax when something should not be enabled, and relying on SFINAE (Substitution Failure Is Not An Error), which ignores the badly formed template function rather than stopping with a compile error. If the expression passed to `std::enable_if` is false, then there is no `::type` field in the generated struct, and so trying to access it will be invalid code. In particular, the `std::enable_if_t` alias has a built in access for the `::type` field.

Based on how that all works, `std::enable_if` disables the function as a whole when the expression is `false`. It's as if the compiler doesn't even see that method for types that produce a `false` expression.

This also means that `std::enable_if` has to be used where a type is expected. Hence people often put it in the template type parameter list, often as some kind of default. It can also be placed on the return value, or on any of the function parameters. It's a bit weird to use on function parameters though, since they vary in number, and there's no natural choice for which parameter to put it around. The return type is a lot more natural since there is always exactly one, even if it's declared to be `void`. Putting it in the template type parameter list is perhaps more natural for some people, since it means the function signature looks normal, but then there are implications when giving explicit type parameters at the point of call.

My personal preference is generally to use the return type.

As a bit of an aside, you can also use the trailing return type syntax with `auto` and `->`, and put `std::enable_if` on a return type listed after the function parameter list.

```cpp
// Without std::enable_if_t
template<typename T>
auto generate(T min, T max) -> T
```

```cpp
// Guarded with std::enable_if_t
template<typename T>
auto generate(T min, T max) -> std::enable_if_t<std::is_arithmetic_v<T>, T>
```